File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/ManifestClient.swift
import Foundation

private let baseChannelKey = "__base__"
private let defaultRuntimeKey = "__default__"

struct ManifestEncryption {
  let alg: String
  let kid: String
  let wrapNonce: String
  let wrappedDek: String
  let nonce: String
}

struct ManifestFileEntry {
  let path: String
  let sha256: String
  let size: Int?
  let url: String
}

struct LatestManifest {
  let version: String
  /// Bundle zip URL. Present for the zip strategy; nil for deltas.
  let url: String?
  /// Zip hash for the zip strategy; canonical filesHash for deltas.
  let sha256: String
  let size: Int
  let runtimeVersion: String?
  let releaseId: String
  let strategy: String
  let forceImmediate: Bool
  let encryption: ManifestEncryption?
  /// Per-file entries for the deltas strategy; nil for zip.
  let files: [ManifestFileEntry]?
}

struct ManifestSignature {
  let kid: String
  let sig: String
  let iat: Int
  let exp: Int
}

enum ManifestClientError: Error {
  case invalidURL
  case invalidResponse
  case requestFailed(String)
  case insecureURL(String)
}

enum ManifestClient {

  static func requireHTTPS(url: URL, allowInsecure: Bool) throws {
    let scheme = url.scheme?.lowercased() ?? ""
    if scheme == "https" { return }
    if allowInsecure {
      let host = url.host?.lowercased() ?? ""
      if host == "localhost" || host == "127.0.0.1" { return }
    }
    throw ManifestClientError.insecureURL(
      "URL must use HTTPS: \(url.absoluteString)"
    )
  }

  static func fetchLatest(
    cdnUrl: String,
    appId: String,
    channel: String?,
    runtimeVersion: String?,
    allowInsecureUrls: Bool = false,
    manifestKeys: [ManifestKey] = []
  ) async throws -> LatestManifest? {
    let sanitizedBase = cdnUrl.replacingOccurrences(
      of: "/+$",
      with: "",
      options: .regularExpression
    )

    let channelKey = channel?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
      ?? baseChannelKey
    let runtimeKey = runtimeVersion?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
      ?? defaultRuntimeKey

    guard let baseURL = URL(string: sanitizedBase) else {
      throw ManifestClientError.invalidURL
    }
    let url = baseURL
      .appendingPathComponent("manifests", isDirectory: true)
      .appendingPathComponent(appId, isDirectory: true)
      .appendingPathComponent(channelKey, isDirectory: true)
      .appendingPathComponent(runtimeKey, isDirectory: true)
      .appendingPathComponent("manifest.json", isDirectory: false)

    try requireHTTPS(url: url, allowInsecure: allowInsecureUrls)

    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.timeoutInterval = 30

    let (data, response) = try await URLSession.shared.data(for: request)
    guard let httpResponse = response as? HTTPURLResponse else {
      throw ManifestClientError.invalidResponse
    }

    if httpResponse.statusCode == 404 || httpResponse.statusCode == 204 {
      return nil
    }

    guard httpResponse.statusCode == 200 else {
      let body = String(data: data, encoding: .utf8) ?? "unknown"
      throw ManifestClientError.requestFailed(
        "HTTP \(httpResponse.statusCode): \(body)"
      )
    }

    guard
      let object = try JSONSerialization.jsonObject(with: data) as? [String: Any],
      let version = object["version"] as? String,
      let sha256 = object["sha256"] as? String,
      let size = object["size"] as? Int
    else {
      throw ManifestClientError.invalidResponse
    }

    let runtimeVersion = (object["runtimeVersion"] as? String)?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .nilIfEmpty

    let signature = parseSignature(object["signature"])
    guard let releaseId = (object["releaseId"] as? String)?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .nilIfEmpty else {
      throw ManifestClientError.invalidResponse
    }

    let strategy = (object["strategy"] as? String)?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .nilIfEmpty ?? "zip"
    let forceImmediate = object["forceImmediate"] as? Bool ?? false
    let encryption = try parseEncryption(object["encryption"])

    let downloadUrl = (object["url"] as? String)?.nilIfEmpty
    var files: [ManifestFileEntry]?

    if strategy == "deltas" {
      files = try parseFiles(object["files"], allowInsecureUrls: allowInsecureUrls)
    } else {
      guard let downloadUrl, let dlURL = URL(string: downloadUrl) else {
        throw ManifestClientError.invalidResponse
      }
      try requireHTTPS(url: dlURL, allowInsecure: allowInsecureUrls)
    }

    if manifestKeys.isEmpty {
      print("[OtaKit] WARNING: No manifest signing keys configured — signature verification is disabled for this request.")
    }

    if !manifestKeys.isEmpty {
      guard let signature else {
        throw ManifestVerifierError.missingSignature
      }

      try ManifestVerifier.verify(
        appId: appId,
        channel: channel,
        version: version,
        sha256: sha256,
        size: size,
        runtimeVersion: runtimeVersion,
        strategy: strategy,
        forceImmediate: forceImmediate,
        encryption: encryption,
        signature: signature,
        trustedKeys: manifestKeys
      )
    }

    return LatestManifest(
      version: version,
      url: downloadUrl,
      sha256: sha256,
      size: size,
      runtimeVersion: runtimeVersion,
      releaseId: releaseId,
      strategy: strategy,
      forceImmediate: forceImmediate,
      encryption: encryption,
      files: files
    )
  }

  private static func parseFiles(
    _ rawValue: Any?,
    allowInsecureUrls: Bool
  ) throws -> [ManifestFileEntry] {
    guard let rawFiles = rawValue as? [[String: Any]], !rawFiles.isEmpty else {
      throw ManifestClientError.invalidResponse
    }

    var entries: [ManifestFileEntry] = []
    entries.reserveCapacity(rawFiles.count)
    for rawFile in rawFiles {
      guard let path = rawFile["path"] as? String,
            let sha256 = rawFile["sha256"] as? String,
            let fileUrl = rawFile["url"] as? String,
            let parsedUrl = URL(string: fileUrl) else {
        throw ManifestClientError.invalidResponse
      }
      try requireHTTPS(url: parsedUrl, allowInsecure: allowInsecureUrls)
      entries.append(
        ManifestFileEntry(
          path: path,
          sha256: sha256,
          size: rawFile["size"] as? Int,
          url: fileUrl
        )
      )
    }
    return entries
  }

  private static func parseEncryption(_ rawValue: Any?) throws -> ManifestEncryption? {
    guard let rawValue, !(rawValue is NSNull) else {
      return nil
    }
    guard let object = rawValue as? [String: Any],
          let alg = object["alg"] as? String,
          let kid = object["kid"] as? String,
          let wrapNonce = object["wrapNonce"] as? String,
          let wrappedDek = object["wrappedDek"] as? String,
          let nonce = object["nonce"] as? String else {
      throw ManifestClientError.invalidResponse
    }
    return ManifestEncryption(
      alg: alg,
      kid: kid,
      wrapNonce: wrapNonce,
      wrappedDek: wrappedDek,
      nonce: nonce
    )
  }

  private static func parseSignature(_ rawValue: Any?) -> ManifestSignature? {
    guard let sigObj = rawValue as? [String: Any],
          let kid = sigObj["kid"] as? String,
          let sig = sigObj["sig"] as? String,
          let iat = sigObj["iat"] as? Int,
          let exp = sigObj["exp"] as? Int else {
      return nil
    }

    return ManifestSignature(kid: kid, sig: sig, iat: iat, exp: exp)
  }
}

private extension String {
  var nilIfEmpty: String? {
    isEmpty ? nil : self
  }
}

File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/ManifestClient.java
package com.otakit.updater;

import android.net.Uri;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import org.json.JSONObject;

final class ManifestClient {

  private static final String BASE_CHANNEL_KEY = "__base__";
  private static final String DEFAULT_RUNTIME_KEY = "__default__";

  static final class ManifestSignature {

    final String kid;
    final String sig;
    final int iat;
    final int exp;

    ManifestSignature(String kid, String sig, int iat, int exp) {
      this.kid = kid;
      this.sig = sig;
      this.iat = iat;
      this.exp = exp;
    }
  }

  static final class LatestManifest {

    final String version;
    final String url;
    final String sha256;
    final int size;
    final String runtimeVersion;
    final String releaseId;

    LatestManifest(
      String version,
      String url,
      String sha256,
      int size,
      String runtimeVersion,
      String releaseId
    ) {
      this.version = version;
      this.url = url;
      this.sha256 = sha256;
      this.size = size;
      this.runtimeVersion = runtimeVersion;
      this.releaseId = releaseId;
    }
  }

  private ManifestClient() {}

  static void requireHTTPS(URL url, boolean allowInsecure) throws Exception {
    String protocol = url.getProtocol().toLowerCase();
    if ("https".equals(protocol)) return;
    if (allowInsecure) {
      String host = url.getHost().toLowerCase();
      if ("localhost".equals(host) || "127.0.0.1".equals(host)) return;
    }
    throw new IllegalStateException("URL must use HTTPS: " + url.toString());
  }

  static LatestManifest fetchLatest(
    String cdnUrl,
    String appId,
    String channel,
    String runtimeVersion,
    boolean allowInsecureUrls,
    java.util.List<ManifestVerifier.KeyEntry> manifestKeys
  ) throws Exception {
    String base = cdnUrl.replaceAll("/+$", "");
    String channelKey =
      channel != null && !channel.trim().isEmpty() ? channel.trim() : BASE_CHANNEL_KEY;
    String runtimeKey =
      runtimeVersion != null && !runtimeVersion.trim().isEmpty()
        ? runtimeVersion.trim()
        : DEFAULT_RUNTIME_KEY;
    Uri baseUri = Uri.parse(base);
    if (baseUri.getScheme() == null || baseUri.getHost() == null) {
      throw new IllegalStateException("Invalid CDN URL");
    }
    Uri urlUri = baseUri
      .buildUpon()
      .appendPath("manifests")
      .appendPath(appId)
      .appendPath(channelKey)
      .appendPath(runtimeKey)
      .appendPath("manifest.json")
      .build();
    URL url = new URL(urlUri.toString());

    requireHTTPS(url, allowInsecureUrls);

    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
    try {
      connection.setRequestMethod("GET");
      connection.setConnectTimeout(15_000);
      connection.setReadTimeout(30_000);

      int status = connection.getResponseCode();
      if (status == 404 || status == 204) {
        return null;
      }
      if (status != 200) {
        String body = readStream(
          connection.getErrorStream() != null
            ? connection.getErrorStream()
            : connection.getInputStream()
        );
        throw new IllegalStateException("Latest request failed (" + status + "): " + body);
      }

      String payload = readStream(connection.getInputStream());
      JSONObject json = new JSONObject(payload);

      String version = json.getString("version");
      String downloadUrl = json.getString("url");
      String sha256 = json.getString("sha256");
      int size = json.getInt("size");

      String responseRuntimeVersion =
        json.has("runtimeVersion") && !json.isNull("runtimeVersion")
          ? json.getString("runtimeVersion").trim()
          : null;
      if (responseRuntimeVersion != null && responseRuntimeVersion.isEmpty()) {
        responseRuntimeVersion = null;
      }

      ManifestSignature signature = parseSignature(json.optJSONObject("signature"));

      String releaseId = null;
      if (json.has("releaseId") && !json.isNull("releaseId")) {
        releaseId = json.getString("releaseId").trim();
      }
      if (releaseId != null && releaseId.isEmpty()) {
        releaseId = null;
      }
      if (releaseId == null) {
        throw new IllegalStateException("Manifest response missing required releaseId");
      }

      requireHTTPS(new URL(downloadUrl), allowInsecureUrls);

      if (manifestKeys == null || manifestKeys.isEmpty()) {
        android.util.Log.w(
          "OtaKit",
          "No manifest signing keys configured — signature verification is disabled for this request."
        );
      }

      if (manifestKeys != null && !manifestKeys.isEmpty()) {
        if (signature == null) {
          throw new IllegalStateException(
            "Manifest signature missing but signing keys are configured"
          );
        }

        ManifestVerifier.verify(
          appId,
          channel,
          version,
          sha256,
          size,
          responseRuntimeVersion,
          signature,
          manifestKeys
        );
      }

      return new LatestManifest(
        version,
        downloadUrl,
        sha256,
        size,
        responseRuntimeVersion,
        releaseId
      );
    } finally {
      connection.disconnect();
    }
  }

  private static String readStream(InputStream input) throws Exception {
    if (input == null) {
      return "";
    }
    try (InputStream stream = input; ByteArrayOutputStream out = new ByteArrayOutputStream()) {
      byte[] buffer = new byte[8192];
      int read;
      while ((read = stream.read(buffer)) > 0) {
        out.write(buffer, 0, read);
      }
      return new String(out.toByteArray(), StandardCharsets.UTF_8);
    }
  }

  private static ManifestSignature parseSignature(JSONObject sigObj) {
    if (sigObj == null) {
      return null;
    }
    if (!sigObj.has("kid") || !sigObj.has("sig") || !sigObj.has("iat") || !sigObj.has("exp")) {
      return null;
    }
    try {
      return new ManifestSignature(
        sigObj.getString("kid"),
        sigObj.getString("sig"),
        sigObj.getInt("iat"),
        sigObj.getInt("exp")
      );
    } catch (org.json.JSONException e) {
      return null;
    }
  }
}
